proposals/0010-validating-resource-container-elements.md

```diff
@@ -0,0 +1,142 @@
+* Impacted Project(s): (LLVM)
+
+*During the review process, add the following fields as needed:*
+
+* PRs: [#NNNN](https://github.com/microsoft/DirectXShaderCompiler/pull/NNNN)
+* Issues: [#75676](https://github.com/llvm/llvm-project/issues/75676)
+
+## Introduction
+Resources are often used in HLSL, with various resource element types (RETs).
+For example:
+```
+RWBuffer<float> rwbuf: register(u0);
+```
+In this code, the RET is `float`, and the resource type is `RWBuffer`. The
+resource type is not a `RawBuffer` variant, and so there is a distinct set
+of rules that define valid RETs for this resource type.
+
+RETs for non-`RawBuffer` variants may include basic types (ints and uints of sizes 16
+and 32, as well as half, and float). Structs that contain fields of these basic
+types (where all fields in the struct have the same type) may also be RETs. 
+Structs that either have structs as fields or arrays of structs as fields may also be
+allowed, as long as everything can fit in 4 32-bit quantities. Additionally, resource 
+types are not allowed within an RET, even if the underlying resource type has a 
+primitive RET.
+
+RETs for `RawBuffer` variants are much less constrained, the only rule is that the RET
+may not be an incomplete type (a handle type or a resource type).
+
+If someone writes `RWBuffer<MyCustomType>` and MyCustomType is not a 
+valid RET, there there should be infrastructure to reject this RET and emit a message 
+explaining why it was rejected as an RET.
+
+## Motivation
+
+Currently, in `clang\lib\CodeGen\CGHLSLRuntime.cpp`, under `calculateElementType` 
+there is a whitelist of valid RETs. 
+Anything that is not an int, uint, nor a floating-point type, will be rejected.
+The whitelist isn't broad enough, because it doesn't include the case where the RET 
+is user-defined. Ideally, a user should be able to determine how any user-defined 
+structure is invalid as an RET. Some system should be in place to more completely 
+enforce the rules for valid and invalid RETs, as well as provide useful information
+on why they are invalid.
+
+For example, `RWBuffer<bool> b : register(u4);` will not emit any error in DXC, 
+but will in clang-dxc, despite the fact that `bool` is a valid RET.
+
+## Proposed solution
+
+The proposed solution is to use some type_traits defined in the std library, create
+some custom type_traits that aren't defined there, and join them together to define a 
+set of conceptual constraints for any RET that is used. These conceptual constraints
+will be applied to every non-`RawBuffer` resource type that is defined, so that all
+non-`RawBuffer` HLSL resources have the same rules about which RETs are valid. 
+Validation will occur upon resource type instantiation. Additionally, certain 
+resource types are `RawBuffer` variants, such as `StructuredBuffer`. Such resource 
+types will have a `[[hlsl::raw_buffer]]` attribute in the attributed type. If this is
+detected, the rules for valid RETs will be loosened.

```

Comment:
I get from this that there will be one set of type_traits for all that will be the most restrictive, so the typed buffer requirements and then the structured buffers will have some kind of exceptions? Is that simple than giving each its own set of type_traits or have I misunderstood?

From the below, it sounds like there will be two sets of type_traits, but this paragraph makes it sound like there will only be one set of type_traits that will be applied to "any RET".

Reply:
Yes, what I intend to communicate is that that `RawBuffer` set of type traits will be applied to those resource types on declaration in HLSLExternalSemaSource.cpp, and the non-`RawBuffer` type traits will be applied to their respective resource types. I'll tweak the wording.